src/main/java/com/google/sps/servlets/EnterQueue.java

```diff
@@ -46,19 +46,30 @@ public void doPost(HttpServletRequest request, HttpServletResponse response) thr
 
     try {
       String classCode = request.getParameter("classCode").trim();
-      Key classKey = KeyFactory.stringToKey(classCode);
-      Entity classEntity = datastore.get(classKey);
 
-      String idToken = request.getParameter("idToken");
-      FirebaseToken decodedToken = authInstance.verifyIdToken(idToken);
+      if (request.getParameter("enterTA") == null) {
+        Key classKey = KeyFactory.stringToKey(classCode);
+        Entity classEntity = datastore.get(classKey);
 
-      ArrayList<String> updatedQueue = (ArrayList) classEntity.getProperty("studentQueue");
-      updatedQueue.add(decodedToken.getUid());
-      classEntity.setProperty("studentQueue", updatedQueue);
+        String idToken = request.getParameter("idToken");
+        FirebaseToken decodedToken = authInstance.verifyIdToken(idToken);
 
-      datastore.put(classEntity);
+        ArrayList<String> updatedQueue = (ArrayList) classEntity.getProperty("studentQueue");
+        String uID = decodedToken.getUid();
+        if (!updatedQueue.contains(uID)) {
+          updatedQueue.add(uID);
+          classEntity.setProperty("studentQueue", updatedQueue);
 
-      if (request.getParameter("enterTA") == null) {
+          datastore.put(classEntity);
+
+          String visitCode = (String) classEntity.getProperty("visitKey");
+          Entity visitEntity = datastore.get(KeyFactory.stringToKey(visitCode));
+
+          Long numVisits = (Long) visitEntity.getProperty("numVisits");
+          visitEntity.setProperty("numVisits", numVisits + 1);
+
+          datastore.put(visitEntity);
+        }
         response.sendRedirect("/queue/student.html?classCode=" + classCode);
       } else {
         response.sendRedirect("/queue/ta.html?classCode=" + classCode);
```

Comment:
Should you verify that the uid is a ta for this class here?

Reply:
Waiting to merge Maryam's PR before I do this!